### DISCUSSION TITLE
整理脚本工具

### DISCUSSION THREAD

**Discussion by JouSF-1409** [seismo-learn/maintenance#43] [General]
数据处理时，经常遇到不同软件数据交互的问题，这时候一个常见的做法是先保存一个多列的文本文件，然后用脚本进行处理。

一些经常复用的功能就可能被拿出来形成单独的脚本。如果功能写得不够好，比如参数，路径写了很多在脚本里面，该怎么梳理，以做到：

1. 代码始终最新
2. 能在不同的项目文件夹内运行。

这时候脚本该怎么整理归纳，运行逻辑该怎么梳理？
尝试过几种不同的方式：

1. 在$HOME/src/ 中，export 路径到PATH中（ 问题：污染PATH与自动补全，为不同方法写的脚本名称混杂，可能造成误用。同时如果是拷贝的脚本，需要对脚本文件依次做chmod，不然就得在脚本里写明perl  xxx.pl, sh xxx.sh，这样就涉及到路径问题。
2. 在~/.bashrc, ~/.zshrc 里将路径名`alias`或 `export`为一个较短的形式，但这样常用的补全插件如YCM, vim-snippet就不太支持补全。
4. 在project/src, project/util 中（ 问题：在同一方法的不同项目里对逻辑微调时，很难做到保持脚本文件最新
5. 按功能划分，并以大写字母开头以示区分，如project/util/util_plot/C_by_rayp.gmt, project/util/util_cals/A_sep.pl， 但还是有一个路径的问题

在逻辑的整理上，由于我平常处理的数据是这样整理的：

step_1/event_1/traces.......
             event_2/traces.......
step2/station_1/traces......
           station_2/traces......

脚本就很自然地写成：
````bash
dir=$1
cd $dir
.......
````
然后在外部使用for 调用，但是这样就涉及到一个脚本相互调用的问题，如果需要用其他脚本生成一些中间的文件，该怎么设计。

**Comment by seisman** [comment]
我目前更倾向于将这些功能写成很多函数而非一堆脚本。

以 Bash 为例，将不同的功能写成函数，保存在一个文件里，例如 `~/src/seisfunc.sh`，其内容如下:
```
process1 () {
   echo "Hello $1 $2"
}

process2 () {
   echo "World $1 $2"
}
```
当在某个项目中需要使用这些功能时，可以写一个脚本，其内容为：
```
source ~/src/seisfunc.sh

cd dir
process1 par1 par2
process2 par3 par4
cd ..
```
这样做的另一个好处在于，process 函数里不用再切换目录，可以避免不同项目目录结构不同的问题。

**Comment by core-man** [comment]
seisman 的推荐方式也是我常用的一种方式。

将路径或目录相关信息通过调用方式传递给脚本或函数，然后使用一个“总”脚本调用其他脚本或函数，这是个总脚本里设置路径或目录。这样做的第三个好处传递的参数有记录，方便日后查阅。

**Comment by JouSF-1409** [comment]
休息的时候正好在整理脚本，大约有了点思路，如果以可读(一个文件内风格一致.....)，易改(函数的大小，变量数量等)为目标，

我是否可以采用一个接近C/Cpp头文件的形式，建立自己的代码库：

1. 先建立一个脚本/函数库（这里的设置比较随意，按功能，按项目都可以，或者类似matlab，一个函数一个文件
4. 然后脚本库下按不同目的与功能，建立不同的中间（头）文件。根据库内的相对路径，`source`/`#include ` 需要的脚本，导入所需的函数
5. 在实际处理的过程中`source` 第二步的脚本（即头文件），这样就能直接使用库里所有的函数了。

这样子可能会比较卡/启动慢，但胜在不会因为老是改参数使得git status 发生变化。

这时候两个问题似乎也能简单解决：

Q1: 每次启动都要合成一遍文本，
S1: 先合并成一个文件（编译），然后包括这个文件

Q2: 包含可执行文件和perl脚本怎么办
S2: 函数中统一使用一个变量表示脚本库的根路径，然后在合并文件的时候获取他。
